### ISSUE TITLE
any way to profile memory usage in dlio?

### ISSUE THREAD

**Issue by krehm** [argonne-lcf/dlio_benchmark#141]
Is there any way to track memory usage in a DLIO benchmark process?   

I ask because I have a test node with 256 GB of memory.   When running unet3d with 9375 files, I see the MPI process grow from 5.5 GB to 11 GB relatively quickly.   Each of the 4 spawn reader processes is 1.5 GB.  That works out to about 18 GB per MPI process.   I can run 12 MPI processes with DAOS and get full accelerator efficiency, but I can't scale up farther because the machine runs out of memory, and processes are too big to swap, I have to reboot the node to get it working again.

My thought was to try to track down what causes the jumps in memory size in the MPI process, maybe there is a way to reduce the amount of RSS space consumed, but if this has been done before, I'd rather learn from the experts.  :-)

**Comment by hariharan-devarajan** [comment]
Can u disable persistent workers within the data loader to false. I was seeing such memory increases with the persistent workers as pytorch keeps spawning more workers while keeping previous workers alive

**Comment by krehm** [comment]
So the memory usage is caused by the pin_memory=True parameter in the DataLoader() instantiation in torch_data_loader.py.  If changed to False, memory usage drops from 11+GB to 2.7 GB.  Individual steps are a tad slower, from 1.36 seconds up to 1.40 seconds.  My config has 4 reader_threads, there are 2 samplers per thread, so pytorch is adding 1 GB of memory to the process for each reader_thread/sampler combo.  

Any idea what is going on when memory is pinned, what is different about the behavior versus when not pinned?   Why is 1 GB being pinned per reader_thread/sampler combo when the tensors are only a few Kbytes in size?   I'd like to understand this better.

**Comment by hariharan-devarajan** [comment]
My understanding the persistent workers + memory pinning was screwing things up. 

With the new changes I recently did, it might have been fixed.

**Comment by krehm** [comment]
Nope, I just upgraded to today's main branch and reran with top (shift M) running in a separate window.  Parent process RES value is 11.2 Gbytes, SHR is 9.5 Gbytes, VIRT is 22.0 Gbytes.
```
PID     USER     PR  NI  VIRT    RES    SHR    S %CPU  %MEM  TIME+     COMMAND
135981  root     20  0   22.0g   11.2g  9.5g   S 8.0   4.5   0:38.18   python3
136124  root     20  0   12.0g   3.2g   1.0g   R 56.1  1.3   1:20.03   python3
136187  root     20  0   11.6g   2.8g   597736 R 14.0  1.1   1:22.17   python3
135998  root     20  0   11.5g   2.7g   597644 S 8.3   1.1   1:21.67   python3
136061  root     20  0   11.5g   2.7g   597644 S 43.5  1.1   1:23.41   python3
```

**Comment by krehm** [comment]
The memory of the parent process starts to grow immediately as it walks through the 4 child processes, 2 samplers each, sending the "start sending tensors" message on the sockets.   By the time all 8 messages have been sent, the parent process is already near the 11 GB size.   This is in the very first epoch, so I don't think persistent workers has any effect here, the problem occurs with the first set of workers.

I will mention that my test node does happen to have a GPU in it.   The torch code uses both pin_memory=True plus it searches for an attached GPU, and if both are true, then pin_memory is activated.   I mention that in case your test machines do NOT have GPUs in them, I don't think pin_memory would have an effect in that case even if pin_memory=True.

**Comment by krehm** [comment]
The pin_memory call is being made in pin_memory() in python3.9/site-packages/torch/utils/data/_utils/pin_memory.py at the first parameter check, in which device is None:
```
    if isinstance(data, torch.Tensor):
        return data.pin_memory(device)
```
I can't step into the data.pin_memory() code because it's C++, but I can see memory allocation occurring there.  At the moment I haven't figured out how to dig deeper, see the actual memory allocations, why they happen.  Of course I expect some allocation, incoming samples are copied to a Queue() in pinned memory, but I don't see why 8 GB would be allocated given that each step contains only 4 samples.   The main process grows to 11 Gbytes for RES quickly and then stays steady after that for all subsequent epochs, so there is not a continuous memory leak.

**Comment by krehm** [comment]
I changed the number of reader_threads from 4 to 8.   In this case the parent process RES size jumped from 11 GB to 19.2 GB, which is about another 8 GB, so the expansion size of the main process appears to go up by 1 GB for each reader_thread/sampler combo.  Does each reader_thread/sampler combo get a separate Queue() of pinned memory, sized at 1 GB each?

**Comment by krehm** [comment]
@hariharan-devarajan not sure, is a name-ping required for a person to see additions to a ticket, other than the creator?

Anyway, I would be interested in your thoughts on the above.  I am running the unet3d benchmark, with num_files_train set to 9376, multiprocessing_context is fork.

**Comment by hariharan-devarajan** [comment]
@krehm This seems odd. Couple of things could be happening. But do u think this could be a garbage collection issue? if so, can u do del of the numpy array data once it's returned to free memory reference for garbage collection?

For my testing, I generally use a GPU-based machine itself, but maybe I didn't come across this due to the number of files and we have large memories on our clusters. 

Maybe u can do a Valgrind run with more workers on one process and see if u spot any memory leaks. NOTE: I have never tried it with a Python app, let alone a Pytorch app.

**Comment by hariharan-devarajan** [comment]
Another quick thing to check is if your samples are smaller(say 4096) would the memory consumption be proportional ?

**Comment by krehm** [comment]
I am running a single DLIO accelerator instance.  I reduced the number of files from 9375 to 168, it made no difference.  By step 5 in the first epoch, the main process is already 11.1 GB in size.  The child processes are 2.7 GB, which is what the parent was just before it started talking to the child processes.  Throughout the rest of the benchmark, the main process does not grow further in size, so it is not a simple continuous memory leak.

I thought it might have something to do with the pytorch getCachingHostAllocator used with cuda, just a wild guess.  I tried setting PYTORCH_NO_CUDA_MEMORY_CACHING=1 but it made no difference.

**Comment by hariharan-devarajan** [comment]
Another hunch. What is the sample size? The default sample size is 200MB, right? Also, is there any prefetching cache setup? If so, what is its value? 

A couple of things to try again.

1. Make smaller samples. It will reduce memory, hopefully.
2. Set the prefetch cache to 0 to force no memory caching for the data loader.

**Comment by hariharan-devarajan** [comment]
Is the 1GB also corresponds to read threads u use? What is the relation between number of read threads and memory. 

I know prefetch factor is not per process but per read thread

**Comment by krehm** [comment]
If I double the number of reader threads, the growth doubles.   Or if I double the prefetch_factor, the growth doubles.   So the growth corresponds to the number of samples in the pin_memory queue at any point in time.

**Comment by hariharan-devarajan** [comment]
As far as I know there is no way to set the prefetch cache size in bytes. I
will try and look the documentation.
